File: camaleonte/camaleonte/python-cc/src/mediums/filesystem.py
import math, time
from abc import ABC, abstractmethod
from enum import Enum
import math
from src.utils import set_hash_byte, get_hash_byte


# will poll sync file at max 60 times per second
POLL_SYNC_FILE_PERIOD = 1/60


class Signal(Enum):
    CLEAR = 0
    ACK = 1
    NACK = 2
    DONE = 3


class Filesystem(ABC):
    def __init__(self) -> None:
        self.channel_pos = -1
        self.client_count = 0 # this is used to optmize VFS calculation
        self.config_file = self.get_all_files()[0]

    def get_files(self) -> list[str]:
        self.update_virtual_filesystem()
        return self.virtual_filesystem[1::]

    def get_client_count(self) -> int:
        config_data = self.read_content(self.config_file)
        return get_hash_byte(config_data)

    def set_client_count(self, cnt: int) -> None:
        config_data = self.read_content(self.config_file)
        modified_config = set_hash_byte(config_data, cnt)
        self.write_content(self.config_file, modified_config)

    def set_channel_pos(self, pos: int) -> None:
        self.channel_pos = pos

    def update_virtual_filesystem(self) -> None:
        if self.channel_pos == -1:
            raise Exception("Didn't connect or wait for connection!")
        # only update if client count changed
        new_client_cnt = self.get_client_count()
        if self.client_count  == new_client_cnt:
            return
        self.client_count = new_client_cnt
        # calculate upper and lower bounds using geometric sequence formula
        all_files = self.get_all_files()
        base = 1
        ratio = 2
        if self.client_count <= base:
            max_clients = 1
        else:
            exponent = math.ceil(math.log(self.client_count/base,ratio))
            max_clients = base * (ratio**exponent)
        files_per_client = (len(all_files)-1) // max_clients
        start_index = self.channel_pos * files_per_client + 1
        # set the virtual_filesystem using the calculated bounds
        self.virtual_filesystem = all_files[start_index:start_index+files_per_client]
        # select the sync file as the first file in the vfs
        self.sync_file = self.virtual_filesystem[0]
        # set the signal of sync file to clear to avoid unintential read/write
        self.set_signal(Signal.CLEAR)
        print("SYNC FILE: ", self.sync_file, start_index, start_index+files_per_client)  # DEBUG

    # Abstract interface
    @abstractmethod
    def get_all_files(self) -> list[str]: pass

    @abstractmethod
    def set_signal(self) -> Signal:
        self.update_virtual_filesystem()
        pass

    @abstractmethod
    def read_signal(self) -> Signal:
        time.sleep(POLL_SYNC_FILE_PERIOD)
        self.update_virtual_filesystem()
        pass

    @abstractmethod
    def write_content(self, file: str, data: bytes) -> None: pass

    @abstractmethod
    def read_content(self, file: str) -> bytes: pass


class HashEncoding(Filesystem):
    pass


class MetadataEncoding(Filesystem):
    @property
    @abstractmethod
    def PROPERTY_SIZE(self) -> int: pass

    @property
    @abstractmethod
    def PROPERTY_COUNT(self) -> int: pass

    @abstractmethod
    def write_properties(self, file: str, properties: dict) -> None: pass

    @abstractmethod
    def read_properties(self, file: str) -> dict: pass

File: camaleonte/camaleonte/metasploit/camaleonte/util/mediums/filesystem.rb
require_relative '../utils'

# enum like Signal object
module Signal
  CLEAR = 0
  ACK = 1
  NACK = 2
  DONE = 3

  VALUES = {
    clear: CLEAR,
    ack: ACK,
    nack: NACK,
    done: DONE
  }.freeze
  NAMES = VALUES.invert.freeze

  # converts numeric int value to signal value
  # (mostly pointless except any non signal number returns as CLEAR)
  def self.from_i(val)
    if val.between?(0, 3)
      return val
    else
      return 0 # default to returning clear
    end
  end

  # converts int value to string representation of signal (uppercase)
  def self.to_s(val)
    name = NAMES[val]
    return 'CLEAR' unless name # default to returning clear

    name.to_s.upcase
  end

  # converts string representation of signal to int value (for e.g. "CLEAR" -> 0)
  def self.from_s(val)
    return CLEAR unless val.is_a?(String) || val.is_a?(Symbol)

    key = val.to_s.downcase.to_sym
    VALUES[key] || CLEAR # default to returning clear
  end
end

# Abstract Filesystem base
module Filesystem
  attr_accessor :channel_pos, :config_file, :virtual_filesystem, :sync_file

  def initialize
    @channel_pos = -1
    @client_count = 0
    @config_file = get_all_files[0]
  end

  # VIRTUAL FILESYSTEM METHODS
  def get_files
    update_virtual_filesystem
    @virtual_filesystem[1..] || []
  end

  def get_client_count
    config_data = read_content(@config_file)
    get_hash_byte(config_data)
  end

  def set_client_count(cnt)
    config_data = read_content(@config_file)
    modified_config = set_hash_byte(config_data, cnt)
    write_content(@config_file, modified_config)
  end

  def set_channel_pos(pos)
    @channel_pos = pos
  end

  def update_virtual_filesystem
    raise 'Didn’t connect or wait for connection!' if @channel_pos == -1

    # only update if client count changes
    new_client_count = get_client_count
    if @client_count == new_client_count
      return
    end

    @client_count = new_client_count

    # calculate upper and lower bounds
    all_files = get_all_files
    a = 1
    r = 2
    geometric_sequence = if @client_count <= a
                           1
                         else
                           exponent = (Math.log(@client_count.to_f / a) / Math.log(r)).ceil
                           a * (r**exponent)
                         end

    files_per_vfs = (all_files.size - 1) / geometric_sequence
    files_per_vfs = files_per_vfs.floor
    start_index = @channel_pos * files_per_vfs + 1
    # puts "#{files_per_vfs}, #{start_index}, #{@channel_pos}" # DEBUG

    @virtual_filesystem = all_files[start_index...(start_index + files_per_vfs)]
    @sync_file = @virtual_filesystem[0]
    set_signal(Signal::CLEAR)
    # puts "SYNC FILE: #{@sync_file}" # DEBUG
  end

  # ABSTRACT METHODS (raise NotImplementedError)
  def get_all_files
    raise NotImplementedError, "#{self.class} must implement get_all_files"
  end

  def set_signal(_sig)
    # update_virtual_filesystem
    raise NotImplementedError, "#{self.class} must implement set_signal"
  end

  def read_signal
    # sleep POLL_SYNC_FILE_PERIOD
    # update_virtual_filesystem
    raise NotImplementedError, "#{self.class} must implement read_signal"
  end

  def write_content(_file, _data)
    raise NotImplementedError, "#{self.class} must implement write_content"
  end

  def read_content(_file)
    raise NotImplementedError, "#{self.class} must implement read_content"
  end
end

# Hash-based encoder inherits from Filesystem
module HashEncoding
  include Filesystem
  # implement protocol-specific methods later
end

# Metadata-based encoder
module MetadataEncoding
  include Filesystem
  def property_size
    raise NotImplementedError, "#{self.class} must define PROPERTY_SIZE"
  end

  def property_count
    raise NotImplementedError, "#{self.class} must define PROPERTY_COUNT"
  end

  def write_properties(_file, _props)
    raise NotImplementedError, "#{self.class} must implement write_properties"
  end

  def read_properties(_file)
    raise NotImplementedError, "#{self.class} must implement read_properties"
  end
end
